### PR TITLE
fix Issue 18729 - dmd -run executes in different environment

### PR DESCRIPTION
**Pull request by marler8997**
TLDR: the following fails on windows
```
echo void main(){} > voidmain.d
dmd -run rdmd.d -m64 voidmain.d

OPTLINK (R) for Win32  Release 8.00.17
Copyright (C) Digital Mars 1989-2013  All rights reserved.
http://www.digitalmars.com/ctg/optlink.html
OPTLINK : Warning 9: Unknown Option : OUT
OPTLINK : Warning 9: Unknown Option : LIBPATH
OPTLINK : Warning 9: Unknown Option : LIBPATH
OPTLINK : Warning 9: Unknown Option : LIBPATH
...
```

I ran into this problem when I was running a D program via `dmd -run` that itself invoked DMD.  The resulting program was invoking the compiler with `-m64`, but the original `dmd -run` did not use `-m64`.  This caused a problem because the first invocation of the compiler set the `LINKCMD` environment variable to the OPTLINK executable, which caused the second invocation to use the OPTLINK executable but it treated it like the MSVC linker because it was running in 64-bit mode.

The solution here is to maintain the original environment (the environment before it was changed by the compiler) when `dmd -run` executes the compiled program.  This PR accomplishes this by saving the original values of environment variables when they are overwritten and then restoring these values before executing the compiled program.

### CODE REVIEW

File: test/runnable/sameenv.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+$DMD -m${MODEL} -of${OUTPUT_BASE}/printenv${EXE} ${EXTRA_FILES}/printenv.d
+${OUTPUT_BASE}/printenv${EXE} > ${OUTPUT_BASE}/envFromExe.txt
+
+$DMD -m${MODEL} -run ${EXTRA_FILES}/printenv.d > ${OUTPUT_BASE}/envFromRun.txt
+
+diff ${OUTPUT_BASE}/envFromExe.txt ${OUTPUT_BASE}/envFromRun.txt
```
Comment (wilzbach) [inline test/runnable/sameenv.sh]: I always found adding `-p` more helpful when looking at errors. Maybe we should overwrite add an `alias diff = diff -p` to the "bash runtime environment"?

Comment (wilzbach) [reply test/runnable/sameenv.sh]: Also maybe you want to compare it with the output from `env`?

Comment (marler8997) [reply test/runnable/sameenv.sh]: output from env might be different, it sorts differently I believe.

Comment (wilzbach) [reply test/runnable/sameenv.sh]: Then normalize both by  `| sort`?

Comment (marler8997) [reply test/runnable/sameenv.sh]: I get different outputs from `env` and this program on windows.  It sorts a bit different.  In any case, we don't really care that the output exactly matches the output of `env`, we just want to make sure running the executable directly matches `dmd -run`.

File: src/dmd/dinifile.d [hunk 2]
```diff
@@ -153,7 +153,7 @@ void updateRealEnvironment(StringTable* environment)
         memcpy(s + namelen + 1, value, valuelen);
         s[namelen + 1 + valuelen] = 0;
         //printf("envput('%s')\n", s);
-        putenv(s);
```
Comment (marler8997) [inline src/dmd/dinifile.d]: TODO: Not sure why this call to `putenv` exists...we should determine why this was put here.

File: src/dmd/link.d [hunk 3]
```diff
@@ -151,6 +172,37 @@ version (Posix)
...
+
+    /**
+    Access the saved environment variable value.
+    */
+    string saved() const { return _saved; }
+
+    /**
+    Call before the overriding the environment variable to save and restore it later.
+    */
+    void saveToRestoreLater()
```
Comment (wilzbach) [inline src/dmd/link.d]: How about wrapping `EnvRestore` as a voldemort type in a function and thus doing this in at construction?

File: src/dmd/link.d [hunk 4]
```diff
@@ -151,6 +172,37 @@ version (Posix)
     }
 }
 
+/**
+Use to restore an environment variable when this structure goes out of scope.
+*/
+struct EnvRestore(string varName)
+{
+    private string _saved;
+    private bool restore;
```
Comment (wilzbach) [inline src/dmd/link.d]: `isSaved` / `hasSaved`?

File: src/dmd/mars.d [hunk 5]
```diff
@@ -266,7 +266,8 @@ private int tryMain(size_t argc, const(char)** argv)
     sections.push(envsection.ptr);
     parseConfFile(&environment, global.inifilename, inifilepath, inifile.len, inifile.buffer, &sections);
     getenv_setargv(readFromEnv(&environment, "DFLAGS"), &arguments);
-    updateRealEnvironment(&environment);
+    // TODO: not sure why dmd is doing this
+    //updateRealEnvironment(&environment);
```
Comment (wilzbach) [inline src/dmd/mars.d]: Because it has the "great concept" of replacing `DFLAGS` completely with the content of the INI file.
See also: https://github.com/dlang/tools/pull/343

File: src/dmd/link.d [hunk 6]
```diff
@@ -32,6 +32,27 @@ version (Windows) extern (C) int putenv(const char*);
...
+int putenv2(const char[] name, const char[] value)
+{
+    auto fullLength = name.length + 1 + value.length;
+    auto buffer = cast(char*)alloca(fullLength + 1);
+    buffer[0               .. name.length] = name[];
+    buffer[name.length]                    = '=';
+    buffer[name.length + 1 .. fullLength ] = value[];
+    buffer[fullLength] = '\0';
+    return putenv(buffer);
+}
```
Comment (wilzbach) [inline src/dmd/link.d]: Imho `{get,put}env2` is a genuinely bad name for functions for D-ified versions.
Maybe `dgetenv` or simply use overloading?

File: src/dmd/link.d [hunk 7]
```diff
@@ -738,15 +791,16 @@ version (Windows)
...
             message("%s %s", cmd, args);
         if (!global.params.mscoff)
         {
             if ((len = strlen(args)) > 255)
             {
-                char* q = cast(char*)alloca(8 + len + 1);
-                sprintf(q, "_CMDLINE=%s", args);
-                status = putenv(q);
+                restore_CMDLINE.saveToRestoreLater();
+                status = putenv2("_CMDLINE", args);
```
Comment (wilzbach) [inline src/dmd/link.d]: Use D strings -> `args[0 .. len]`

File: src/dmd/link.d [hunk 8]
```diff
@@ -32,6 +32,27 @@ version (Windows) extern (C) int putenv(const char*);
 version (Windows) extern (C) int spawnlp(int, const char*, const char*, const char*, const char*);
 version (Windows) extern (C) int spawnl(int, const char*, const char*, const char*, const char*);
 version (Windows) extern (C) int spawnv(int, const char*, const char**);
+
+string getenv2(const(char)* name)
+{
+    auto result = getenv(name);
+    return cast(string)result[0 .. strlen(result)];
+}
+int putenv2(const char[] name, const(char)* value)
```
Comment (wilzbach) [inline src/dmd/link.d]: Do we really need the C-string overload? Remember they "must die" ;-)

File: src/dmd/link.d (hunk 3 above)
Comment (marler8997) [reply src/dmd/link.d]: it doesn't always save the environment variable, only in some cases.

File: src/dmd/globals.d [hunk 9]
```diff
@@ -68,6 +73,75 @@ enum CPU
     native              // the machine the compiler is being run on
 }
 
+/**
+Used to store an environment variable including its value.
+*/
+extern (C++) struct EnvironmentVariable
```
Comment (ibuclaw) [inline src/dmd/globals.d]: Why does this need to be in global? Nobody apart from dmd would use this.

File: src/dmd/globals.d [hunk 10]
```diff
@@ -219,6 +294,7 @@ struct Param
 
     bool run; // run resulting executable
     Strings runargs; // arguments for executable
+    Array!EnvironmentVariable envToRestoreBeforeRun;
```
Comment (ibuclaw) [inline src/dmd/globals.d]: This is only accessed by private functions, and shouldn't be in global params accessible by all.

**Review by ibuclaw — DISMISSED** [review]
For better encapsulation, these routines should be free functions and declared `extern(D)`, if dinifile or utils isn't a good place, then it's own module, where both EnvironmentVariable struct and array are private.

**Comment by marler8997** [comment]
@ibuclaw good suggestions.  Thanks for taking the time to review.

File: src/win32.mak [hunk 11]
```diff
@@ -153,7 +153,7 @@ FRONT_SRCS=$D/access.d $D/aggregate.d $D/aliasthis.d $D/apply.d $D/argtypes.d $D
 	$D/cond.d $D/constfold.d $D/cppmangle.d $D/cppmanglewin.d $D/ctfeexpr.d $D/ctorflow.d $D/dcast.d $D/dclass.d		\
 	$D/declaration.d $D/delegatize.d $D/denum.d $D/dimport.d $D/dinifile.d $D/dinterpret.d	\
 	$D/dmacro.d $D/dmangle.d $D/dmodule.d $D/doc.d $D/dscope.d $D/dstruct.d $D/dsymbol.d $D/dsymbolsem.d		\
-	$D/lambdacomp.d $D/dtemplate.d $D/dversion.d $D/escape.d			\
+	$D/lambdacomp.d $D/dtemplate.d $D/dversion.d $D/env.d $D/escape.d			\
```
Comment (stefan-koch-sociomantic) [inline src/win32.mak]: positional difference when compared to posix makefile.

**Review by stefan-koch-sociomantic — CHANGES_REQUESTED** [review]
Please keep the make files as much as possible in sync

File: src/win32.mak (hunk 11 above)
Comment (wilzbach) [reply src/win32.mak]: @stefan-koch-sociomantic this is fine. The windows Makefile is only semi-ordered anyhow ...

**Comment by marler8997** [comment]
Ping

**Comment by marler8997** [comment]
Ping

**Comment by marler8997** [comment]
Re-opening, I'm hopeful that with the re-organization, that this one will be able to get in.

**Comment by JinShil** [comment]
Forgive me if I'm missing the point, but shouldn't `dmd -m64 -run rdmd.d -m64 voidmain.d` do the right thing?  That is, the position of the switches relative to the `-run` switch determines which switches are passed to which invocation?

**Comment by marler8997** [comment]
Yeah that command would work. Though, it's different than the problem command that this PR fixes. Dmd uses environment variables to pass settings to other programs, but then doesn't clean up those environment variables before invoking the resulting executable with -run.

**Comment by marler8997** [comment]
By the way, if you're sitting in the dlang tools repo, you can do this:
```bash
./rdmd.d rdmd.d rdmd.d rdmd.d ...[repeat forver]
```
You can just keep going, and eventually it will just print out the rdmd usage...lol :)

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/env.d [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+struct Env
+{
+    private const(char)* setString; /// null-terminated string of the form "VAR=VALUE"
+    private uint length;            /// length of setString
+    private uint equalsIndex;       /// index of '=' in setString
+
+    /**
+    Access the name of the environment variable as a D string.
+    */
+    auto nameString() const { return setString[0 .. equalsIndex]; }
```
Comment (jacob-carlborg) [inline src/dmd/env.d]: Why the `String` suffix?

File: src/dmd/env.d [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+import core.sys.posix.stdlib;
+import dmd.globals;
+import dmd.root.array;
+import dmd.root.rmem;
+
+version (Windows) extern (C) int putenv(const char*) nothrow;
+
+/**
+Used to store an environment variable including its value.
+*/
```
Comment (jacob-carlborg) [inline src/dmd/env.d]: You can save two lines by using the `///` syntax.

File: src/dmd/env.d [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+struct Env
+{
+    private const(char)* setString; /// null-terminated string of the form "VAR=VALUE"
```
Comment (jacob-carlborg) [inline src/dmd/env.d]: Can we use a D string?

Comment (marler8997) [reply src/dmd/env.d]: We could but I don't see the benefit here.

Comment (marler8997) [reply src/dmd/env.d]: Right but this isn't being used as a string.  It's just being passed to the C function `putenv`.

Comment (marler8997) [reply src/dmd/env.d]: I could still make it a string, but this isn't really being used like other strings we want to convert to D strings, hence why I don't see the benefit.  Maybe you can show me a place where using a D string would prevent an error?  I don't see any.

Comment (marler8997) [reply src/dmd/env.d]: This member MUST be a null-terminated string.  And I've noticed that when DMD has to use a null-terminated string, it usually uses `char*` to signify that.  If I use `char[]` would that be confusing?

Comment (marler8997) [reply src/dmd/env.d]: Ok I've noticed other places where DMD uses D arrays for strings even when they need to be null-terminated.  So I've gone ahead and made it a `string`.

**Review by WalterBright — DISMISSED** [review]
First off, bug fixes should have a bugzilla issue filed for them. Second, I'm not sure this is even a bug. Why should running a compiled program inherit the environment set by the compiler? The two have no necessary connection between them. I think it's a bug if it **did** inherit the environment variables set by the compiler. This shouldn't be polymorphism.

I.e. need a much better justification.

**Comment by WalterBright** [comment]
Ok, I see it does have a bugzilla issue. I fixed the title to follow the usual form for bug fixes.

**Comment by marler8997** [comment]
> Why should running a compiled program inherit the environment set by the compiler?

It shouldn't that's the bug.

> The two have no necessary connection between them. I think it's a bug if it did inherit the environment variables set by the compiler. This shouldn't be polymorphism.

Agreed.  You're describing what the compiler is doing today.  This PR fixes it so that the executable is run from the original environment before the compiler changed it during compilation.

**Comment by marler8997** [comment]
To be fair, I read the description and could see how the purpose could be confused.  I updated the description to be more clear.

> The solution here is to maintain the original environment (the environment before it was changed by the compiler) when dmd -run executes the compiled program. This PR accomplishes this by saving the original values of environment variables when they are overwritten and then restoring these values before executing the compiled program.

File: src/dmd/link.d [hunk 15]
```diff
@@ -1053,12 +1048,15 @@ version (Windows)
...
 
-                    char* npath = cast(char*)mem.xmalloc(5 + pathlen + 1 + addpathlen + 1);
+                    const length = 5 + addpathlen + 1 + pathlen;
+                    char* npath = cast(char*)mem.xmalloc(length + 1);
                     memcpy(npath, "PATH=".ptr, 5);
                     memcpy(npath + 5, addpath, addpathlen);
                     npath[5 + addpathlen] = ';';
                     memcpy(npath + 5 + addpathlen + 1, path, pathlen + 1);
-                    putenv(npath);
+                    if (0 != LocalEnvVar(cast(string)npath[0 .. length], 4).putenvRestorable)
```
Comment (thewilsonator) [inline src/dmd/link.d]: please don't use the yodel style

**Comment by thewilsonator** [comment]
Apart from the nit, this looks good.

File: src/dmd/link.d (hunk 15 above)
Comment (marler8997) [reply src/dmd/link.d]: Actually in this cast, `putenvRestorable` is a member of `LocalEnvVar`

Comment (thewilsonator) [reply src/dmd/link.d]: I meant remove the `0 !=`

Comment (marler8997) [reply src/dmd/link.d]: gotcha, done

**Comment by marler8997** [comment]
I emailed Walter and he just got back to me saying he's going to take another look at this PR after I clarified what it was doing.
> Thanks for clearing up my confusion. I'll see about getting to it. -Walter

**Comment by marler8997** [comment]
Still waiting on @WalterBright  to review.  I've emailed him a couple times and he says he hasn't had time to review yet.

In the meantime, since #9848 was merged, I've converted the new BASH test I added in this PR to D.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/env.d [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+*/
+int putenvRestorable(const(char)[] name, const(char)[] value) nothrow
+{
+    auto var = LocalEnvVar.xmalloc(name, value);
+    const result = var.putenvRestorable();
+    var.xfree(result);
+    return result;
+}
+
+/// Holds a `VAR=value` string that can be put into the global environment.
```
Comment (WalterBright) [inline src/dmd/env.d]: It's "name=value".

File: src/dmd/env.d [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+struct LocalEnvVar
+{
+    string nameValueCStr;        // The argument passed to `putenv`.
```
Comment (WalterBright) [inline src/dmd/env.d]: Note it's called "nameValue" here, not "VARvalue"

File: src/dmd/env.d [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+import dmd.root.array;
+import dmd.root.rmem;
+import dmd.utils;
+
+version (Windows)
+    private extern (C) int putenv(const char*) nothrow;
+
+/**
+Construct a variable from `name` and `value` and put it in the environment while saving
+the current value of the environment variable.
```
Comment (WalterBright) [inline src/dmd/env.d]: No `Params:` section. Function does not construct a variable. Should return a bool.

File: src/dmd/env.d [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+Returns:
+    0 on success, non-zero on failure
+*/
+int putenvRestorable(const(char)[] name, const(char)[] value) nothrow
+{
+    auto var = LocalEnvVar.xmalloc(name, value);
+    const result = var.putenvRestorable();
```
Comment (WalterBright) [inline src/dmd/env.d]: Having two different functions with the same name is confusing.

File: src/dmd/env.d [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    auto name() const { return nameValueCStr[0 .. equalsIndex]; }
+
+    /// The value of the variable
+    auto value() const { return nameValueCStr[equalsIndex + 1 .. $]; }
+
+    /**
+    Put this variable in the environment while saving the current value of the
+    environment variable.
+    Returns:
+        0 on success, non-zero on failure
```
Comment (WalterBright) [inline src/dmd/env.d]: use bool

File: src/dmd/env.d [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+    /// The name of the variable
+    auto name() const { return nameValueCStr[0 .. equalsIndex]; }
+
+    /// The value of the variable
+    auto value() const { return nameValueCStr[equalsIndex + 1 .. $]; }
+
+    /**
+    Put this variable in the environment while saving the current value of the
+    environment variable.
```
Comment (WalterBright) [inline src/dmd/env.d]: save it where?

File: src/dmd/env.d [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+        0 on success, non-zero on failure
+    */
+    int putenvRestorable() const nothrow
+    {
+        RestorableEnv.save(name);
+        return .putenv(cast(char*)nameValueCStr.ptr);
+    }
+
+    /**
+    Allocate a new variable via xmalloc that can be promoted to the global environment.
```
Comment (WalterBright) [inline src/dmd/env.d]: `promoted`? more like `added`

File: src/dmd/env.d (hunk 17 above)
Comment (WalterBright) [reply src/dmd/env.d]: Note in the comment that it has a terminating 0 past the end of the string, not as part of the name.

File: src/dmd/env.d [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+        version (Windows)
+            doFree = true;
+        else
+        {
+            // on posix, when putenv succeeds ownership of the memory is transferred
+            // to the global environment
+            doFree = (putenvResult != 0);
+        }
+        if (doFree)
+            mem.xfree(cast(void*)nameValueCStr.ptr);
```
Comment (WalterBright) [inline src/dmd/env.d]: get rid of `doFree` and you can collapse this to 4 lines of code instead of 7.

File: src/dmd/env.d [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+    /// Holds the original values of environment variables when they are overwritten.
+    private __gshared Array!LocalEnvVar originalVars;
+
+    /// Restore the original environment.
+    static void restore()
+    {
+        foreach (var; originalVars)
+        {
+            if (0 != putenv(cast(char*)var.nameValueCStr))
+                assert(0, "putenv failed");
```
Comment (WalterBright) [inline src/dmd/env.d]: Really don't need elaborate assert's in DMD. If people can't find where an assert failed from the stack trace that gives file/line, they shouldn't be working on DMD.

File: src/dmd/env.d [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+{
+    /// Holds the original values of environment variables when they are overwritten.
+    private __gshared Array!LocalEnvVar originalVars;
+
+    /// Restore the original environment.
+    static void restore()
+    {
+        foreach (var; originalVars)
+        {
+            if (0 != putenv(cast(char*)var.nameValueCStr))
```
Comment (WalterBright) [inline src/dmd/env.d]: Not a fan of `0 !=`. We can expect some minimal knowledge of how boolean conditions work in D for people working on DMD.

File: src/dmd/env.d [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+    Allocate a new variable via xmalloc that can be promoted to the global environment.
+    Params:
+        name = name of the variable
+        value = value of the variable
+    Returns:
+        a newly allocated variable that can be promoted to the global environment
+    */
+    static LocalEnvVar xmalloc(const(char)[] name, const(char)[] value) nothrow
```
Comment (WalterBright) [inline src/dmd/env.d]: Calling this `xmalloc` is confusing it with `Mem.xmalloc`. Same with `xfree`.

File: src/dmd/env.d [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+    /// Save the environment variable `name` if not saved already.
+    static void save(const(char)[] name) nothrow
+    {
+        foreach (var; originalVars)
+        {
+            if (name == var.name)
+                return; // already saved
+        }
+        originalVars.push(LocalEnvVar.xmalloc(name,
+            name.toCStringThen!(n => getenv(n.ptr)).toDString));
```
Comment (WalterBright) [inline src/dmd/env.d]: There seems to be a lot of converting `char*` to `[]` and back again. Wouldn't it be simpler to just leave them as `char*`?

Comment (marler8997) [reply src/dmd/env.d]: > There seems to be a lot of converting char* to [] and back again
Alot? This is the one and only place in this change that we're calling `toCStringThen`.

We're using the `char[]` version of `name` for both the comparison and to pass to the `LocalEnvVar` allocation function.  We only need the `char*` version to call `getenv`.  I'm not sure what you'd like me to change here.

**Comment by marler8997** [comment]
@WalterBright I've integrated all your suggestions except the one about `toCStringThen` since I'm not sure what you want for that suggestion.  Please let me know what you'd like me to do next here.

**Comment by WalterBright** [comment]
Thank you. In thinking about it some more, what's happening is simply storing `name=value` pairs, and avoiding making duplicate entries for `name`. This is exactly what D associative arrays do. You can dispense with the new structs, and just use:
```
__gshared string[string] envNameValues;
```
Use the `in` operator to test for membership, and `foreach` to iterate over `envNameValues`.

**Comment by marler8997** [comment]
Ok, I've changed the code per your suggestions.

**Comment by WalterBright** [comment]
Much better.

**Comment by marler8997** [comment]
Thanks alot for seeing this through to the finish line.
